src/haywire/http_server.c

```diff
@@ -378,6 +385,132 @@ void http_stream_on_read_http_parser(uv_stream_t* tcp, ssize_t nread, const uv_b
     }
 }
 
+void http_stream_on_read_pico(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* buf)
+{
+    http_connection* connection = (http_connection*)tcp->data;
+
+    if (nread > 0)
+    {
+        // TODO: Should this be at a higher level? Both parsers call this the same way.
+        http_request_buffer_consume(connection->buffer, nread);
+
+        printf("START\tNREAD: %d BUFUSED: %d PREVBUFLEN: %d\n",
+               nread,
+               http_request_buffer_get_used(connection->buffer),
+               connection->prevbuflen);
+        //fwrite(http_request_buffer_get_buffer(connection->buffer),
+        //       http_request_buffer_get_used(connection->buffer), 1, stdout);
+
+        while (connection->prevbuflen < http_request_buffer_get_used(connection->buffer))
+        {
+            printf("LOOP\tNREAD: %d BUFUSED: %d PREVBUFLEN: %d\n",
+                   nread,
+                   http_request_buffer_get_used(connection->buffer),
+                   connection->prevbuflen);
+
+            int parsed;
+            char buf[4096], *method, *path;
+            int minor_version;
+            struct phr_header headers[100];
+            size_t method_len;
+            size_t path_len;
+            size_t num_headers;
+            num_headers = sizeof(headers) / sizeof(headers[0]);
+
+            parsed = phr_parse_request(&http_request_buffer_get_buffer(connection->buffer)[connection->prevbuflen],
+                                       http_request_buffer_get_used(connection->buffer),
```

Comment:
I don't remember that comment, but I think there are a few issues with the current approach.

1 Looking at signature of `phr_parse_request` into account, `last_len` should tell the parser where
```C
int phr_parse_request(const char *buf_start, size_t len, const char **method, size_t *method_len, const char **path,
                      size_t *path_len, int *minor_version, struct phr_header *headers, size_t *num_headers, size_t last_len)
```
`last_len` is hard-coded to 0. That parameter is used by pico parser to determine if a request has been fully parsed or not. Based on the example at https://github.com/h2o/picohttpparser, it seems like it's supposed to be how far along are we in the buffer after the last round of parsing. Sounds to me like it should really be `http_request_buffer_get_buffer(connection->buffer)->used - nread`, i.e., the bit on the buffer that is yet to be parsed.

2) I don't get the loop in its current form.

`while (connection->prevbuflen < http_request_buffer_get_used(connection->buffer))`

`phr_parse_request` returns a positive number, when it's reached the end of a request. There may be multiple requests in the buffer we're getting, so I think we should loop until all calls from `phr_parser_request` have consumer all `nread` bytes. Example:

```C
int total_parsed = 0;
int parsed;

while (total_parsed < nread) {
  parsed = phr_parser_request(...., total_parsed);
  
  if (parsed > 0) {
     total_parsed += parsed;   
     http_request_buffer_mark(connection->buffer); // so we can sweep the contents of the buffer the next time we get a new chunk of data from libuv
  }
}

http_request_buffer_sweep(connection->buffer); // we've read all there was to read from this buffer, so we can sweep any buffer chunks that are guaranteed not to be in use. 
```

This is as described in:

```c
void http_request_buffer_mark(hw_request_buffer* buf) {
    http_request_buffer* buffer = (http_request_buffer*) buf;
    /* unfortunately, the parser doesn't tell us where the request ends exactly,
     * so the only thing we can be sure is that it ends in the current buffer chunk, so anything before it can
     * effectively be swept, so we're placing the mark at that point now. */
    buffer->mark = buffer->used_before;
}
```

The following example describe how things should be happening for them to work.

Imagine a buffer size of 64KB and two requests of 32KB received 24KB at the time:

1) 64KB are allocated

2) 24KB of request 1 are read into the `hw_request_buffer`. Let's call that chunk 1. There are still 40KB free.

3) The parser parses those 24KB without completing the request. We can't sweep anything from the `hw_request_buffer` buffer as otherwise we'd lose the first 24KB of the first request, so `sweep` is actually a no-op in this case, because a mark hasn't been set yet (i.e. we haven't gotten to the end of a single request so far).

4) We get another 24KB from libuv, on the same physical buffer (hw_request_buffer->buffer), but 24KB from its start. Let's call that chunk 2. 8KB belong to the request 1 and 16KB to request 2.

5) The parser parses 8KB and returns. We place a mark because we now have request 1 in full. This will effectively mark anything before the beginning of chunk #2 as content that can be swept.

6) We still have 16KB to process for the request 2, before processing the next chunk from libuv. We parser parses 16KB, but returns 0 as it hasn't yet finished.

7) We get to the end of the chunk #2 and call sweep. This time, there's a mark, right at the beginning of chunk #2. This will effectively move chunk #2 to the beginning of the physical buffer and overwrite chunk #1.

8) We get another 16KB this time (assume no more requests left), which I'll refer to as chunk #3. The parser runs and returns 16KB. We mark again, therefore saying that anything before the beginning of chunk #3 can be swept now (those first 8KB of the request 1 and 16KB of the request 2).

9) We get to the end of chunk 3, so chunk 2 gets swept away and chunk 3 takes its place in the physical buffer.

I've tried to represent the same sequence in a set of diagrams (sorry, my writing is terrible):

![img_3736 2](https://user-images.githubusercontent.com/2456228/27507886-7aa95668-58d1-11e7-8a9c-289b2c4ba22f.JPG)